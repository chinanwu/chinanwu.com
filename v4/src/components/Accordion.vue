<template>
  <div>
    <button
      id="accordionBtn"
      class="Accordion__btn"
      @click="handleClick"
      @keydown="handleKeyDown"
    >
      <span>{{ label }}</span>
      <span>{{ expand ? "-" : "+" }}</span>
    </button>
    <transition name="expand" mode="out-in">
      <slot v-if="expand" name="content"></slot>
    </transition>
  </div>
</template>

<script>
import validEvent from "@/functions/validEvent";

export default {
  name: "Accordion",
  props: {
    label: {
      type: String,
      default: null
    },
    expand: {
      type: Boolean,
      default: false
    }
  },
  emits: ["change"],
  methods: {
    handleClick() {
      this.$emit("change");
    },
    handleKeyDown(event) {
      if (validEvent(event, ["Enter", "Space"])) {
        event.preventDefault();
        this.handleClick();
      }
    }
  }
};
</script>

<style lang="less" scoped>
@import "../styling/colours";

.Accordion__btn {
  width: 100%;
  text-align: left;
  padding: 1rem;
  background-color: @grey-05;
  border: none;
  display: flex;
  justify-content: space-between;
  cursor: pointer;
  transition: background-color 0.2s;
  border-radius: 0.5rem;
  font-size: 1.8rem;

  &:hover,
  &:focus {
    background-color: @grey-06;
  }

  &:active {
    background-color: @grey-04;
  }
}

.expand {
  &-enter-active,
  &-leave-active {
    transition: all 0.3s;
    overflow: hidden;
  }
  &-enter-from,
  &-leave-to {
    max-height: 0;
    opacity: 0;
  }
  &-enter-to,
  &-leave-from {
    max-height: 100vh;
    opacity: 1;
  }
}
</style>
