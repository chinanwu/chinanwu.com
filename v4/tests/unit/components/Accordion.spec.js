import { shallowMount, mount } from "@vue/test-utils";
import Accordion from "@/components/Accordion.vue";

describe("Accordion", () => {
  describe("rendering", () => {
    it("renders", () => {
      const wrapper = shallowMount(Accordion);
      expect(wrapper.element).toMatchSnapshot();
    });

    it("renders expanded", () => {
      const wrapper = mount(Accordion, {
        slots: {
          content: "<div id='test'>Here</div>"
        },
        props: {
          labelExpanded: "Expanded",
          expand: true
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.find("#test").text()).toEqual("Here");
      expect(wrapper.find("#accordionBtn").text()).toEqual("Expanded-");
    });

    it("renders minimized", () => {
      const wrapper = shallowMount(Accordion, {
        slots: {
          content: "<div id='test'>Here</div>"
        },
        props: {
          labelMinimized: "Minimized"
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.find("#accordionBtn").text()).toEqual("Minimized+");
    });
  });

  describe("behaviour", () => {
    it("emits change when clicked", () => {
      const wrapper = shallowMount(Accordion);
      expect(wrapper.element).toMatchSnapshot();
      wrapper.find("#accordionBtn").trigger("click");
      expect(wrapper.emitted()).toHaveProperty("change");
    });
  });
});
