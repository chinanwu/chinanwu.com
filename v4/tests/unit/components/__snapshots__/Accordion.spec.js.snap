// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion behaviour emits change when clicked 1`] = `
<div>
  <button
    aria-live="polite"
    class="Accordion__btn"
    id="accordionBtn"
  >
    <span />
    <span
      role="presentation"
    >
      +
    </span>
  </button>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
`;

exports[`Accordion rendering renders 1`] = `
<div>
  <button
    aria-live="polite"
    class="Accordion__btn"
    id="accordionBtn"
  >
    <span />
    <span
      role="presentation"
    >
      +
    </span>
  </button>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
`;

exports[`Accordion rendering renders expanded 1`] = `
<div>
  <button
    aria-live="polite"
    class="Accordion__btn"
    id="accordionBtn"
  >
    <span>
      Expanded
    </span>
    <span
      role="presentation"
    >
      -
    </span>
  </button>
  <transition-stub>
    
    
    <div
      id="test"
    >
      Here
    </div>
    
    
  </transition-stub>
</div>
`;

exports[`Accordion rendering renders minimized 1`] = `
<div>
  <button
    aria-live="polite"
    class="Accordion__btn"
    id="accordionBtn"
  >
    <span>
      Minimized
    </span>
    <span
      role="presentation"
    >
      +
    </span>
  </button>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
`;
